src/libopensc/card-starcos.c

```diff
@@ -1728,6 +2003,36 @@ static int starcos_get_serialnr(sc_card_t *card, sc_serial_number_t *serial)
 	}
 
 	switch (card->type) {
+		case SC_CARD_TYPE_STARCOS_V3_2:{
+			size_t taglen = 0;
+			int  offs;
+			const u8 *tag = NULL;
+
+			sc_format_apdu(card, &apdu, SC_APDU_CASE_2_SHORT, 0xCA, 0xDF, 0x20);
+			apdu.cla = 0x00;
+			apdu.resp = rbuf;
+			apdu.resplen = sizeof(rbuf);
+			apdu.le = 256;
+			apdu.lc   = 0;
+			apdu.datalen = 0;
+
+			// Send APDU
+			r = sc_transmit_apdu(card, &apdu);
+			SC_TEST_RET(card->ctx, SC_LOG_DEBUG_NORMAL, r, "APDU transmit failed");
+			if (apdu.sw1 != 0x90 || apdu.sw2 != 0x00)
+				return SC_ERROR_INTERNAL;
+
+			//Serial nr contained in data obj with following tag
+			tag = sc_asn1_find_tag(card->ctx, apdu.resp, apdu.resplen, 0x9F6A, &taglen);
+			if (tag  == NULL) {
+				return SC_ERROR_INTERNAL;
+			}
+			//only consider last 8 bytes (apparently common practice)
+			offs = taglen > 8 ? taglen - 8 : 0;
+			memcpy(&card->serialnr, tag + offs, MIN(8, taglen));
```

Comment:
out of bounds read. should be taglen-offs

Reply:
Here, tag points to the beginning of a tag which is potentially longer than 8 bytes. If this is the case, the offset is specified so that only the last 8 bytes are read out and used for the serial number. Hence, offs = taglength - 8 is added so that tag + offs points to the start of the last 8 bytes.

Reply:
nevermind, that should work as intended.

However, `card->serialnr.len = 8;` needs to be changed to `card->serialnr.len = MIN(8, taglen);`